server/aesdsocket.c

```diff
@@ -37,36 +124,268 @@ static void signal_handler (int signo)
-		close(client_fd);
-	        close(write_fd);
-		closelog();
-		free(buffer);
-		free(transmit_buffer);
-		remove(FILE_PATH);
+		shutdown(socket_fd,SHUT_RDWR);
+		stop = 1;
+		close_all();
+		exit (EXIT_SUCCESS);
 	}
+	
         else
        	{
                 /* this should never happen */
                 printf ("Unexpected signal!\n");
                 exit (EXIT_FAILURE);
         }
 	
-        exit (EXIT_SUCCESS);
+        
+}
+
+void* receive_send_data(void* thread_param)
+{
+	
+	thread_data_t *thread_func_args = (thread_data_t*)thread_param;
+
+	int received_bytes, write_bytes, read_bytes, send_bytes;
+	int loc = 0;
+	int count = 1;
+
+	// allocate memory for buffers
+        thread_func_args->buffer = (char *)malloc(BUFFER_SIZE*sizeof(char));
+        if(thread_func_args->buffer == NULL)
+        {
+                printf("Error while doing malloc\n");
+                close_all();
+                exit (EXIT_FAILURE);
+        }
+
+        thread_func_args->transmit_buffer = (char *)malloc(BUFFER_SIZE*sizeof(char));
+        if(thread_func_args->transmit_buffer == NULL)
+        {
+                printf("Error while doing malloc\n");
+                close_all();
+                exit (EXIT_FAILURE);
+        }
+
+
+
+	//receive data
+	do
+	{
+                received_bytes = recv(thread_func_args->client_fd, thread_func_args->buffer + loc, BUFFER_SIZE, 0); //sizeof(buffer)
+                if(received_bytes <= 0)
+                {
+                        printf("Error while receiving\n");
+                }
+
+                loc += received_bytes;
+
+                // increase buffer size dynamically if required
+                count += 1;
+                thread_func_args->buffer = (char *)realloc(thread_func_args->buffer, count*BUFFER_SIZE*(sizeof(char)));
+                if(thread_func_args->buffer == NULL)
+                {
+                        close(thread_func_args->client_fd);
+                        close_all();
+                        printf("Error while doing realloc\n");
+                        exit (EXIT_FAILURE);
+                }
+
+
+	}while(strchr(thread_func_args->buffer, '\n') == NULL);
+	received_bytes = loc;
+	thread_func_args->buffer[received_bytes] = '\0';
+	//total_bytes += received_bytes; //keep track of total bytes
+
+	int rc;
+	rc = pthread_mutex_lock(thread_func_args->mutex); //lock mutex
+	if(rc != 0)
+	{
+		perror("Mutex lock failed\n");
+		close_all();
+		exit (EXIT_FAILURE);
+	} 	
+					
+	// block the signals SIGINT and SIGTERM
+	rc = sigprocmask(SIG_BLOCK, &thread_func_args->signal_mask, NULL);
+	if(rc == -1)
+	{
+		printf("Error while blocking the signals SIGINT, SIGTERM");
+		close_all();
+		exit (EXIT_FAILURE);							
+	}			
+					
+	// write to the file
+	write_bytes = write(write_fd, thread_func_args->buffer, received_bytes);
+	if(write_bytes < 0)
+	{
+		printf("Error in writing bytes\n");
+		close(thread_func_args->client_fd);
+		close_all();
+		exit (EXIT_FAILURE);
```

Comment:
@midhun9 I'm not sure you need to kill the entire process in each of these error cases.  What would happen with your mutex if you changed one of these to `return` instead?